### DISCUSSION TITLE
Editable edge paths

### DISCUSSION THREAD

**Discussion by cpotdevin** [xyflow/xyflow#1029] [New Features]
Implement an `editable` edge type with the ability to click and drag to generate corners on the edge's path. This would help to organize edges in large diagrams.

Currently I am working on a visualization tool for digital circuit teaching and learning. The react-flow library has been **extremely** useful. In the tool diagrams can start to get quite complex once you have a couple of nodes with multiple inputs and outputs. By being able to tweak the path of nodes the user could be able to clean up the diagram and make it more readable. I would be open to help bring this functionality to fruition if there is any interest.

**Comment by ottovw** [comment]
Just created an issue for this, as I wasn't aware of the discussion page.

Absolutely agree, this would be great addition. We called it "Mid Points". Here's a picture to illustrate the feature request:
![image](https://user-images.githubusercontent.com/1045946/139501251-facd9b5d-e352-4952-926d-2e8bb91dfc53.png)

**Comment by Keanottesen** [comment]
I have been trying to create this as well. Would really appreciate if it was a part of the library

**Comment by tisoap** [comment]
For anyone that wants an Edge today that does not intersect with other nodes, I've created an implementation that uses path-finding to achieve that:
[@tisoap/react-flow-smart-edge](https://www.npmjs.com/package/@tisoap/react-flow-smart-edge)

**Comment by ghost** [comment]
currently stuck on the same problem... are there any news on this topic? @moklick

**Reply by MaxKoldun** [reply]
@konsti-papa @moklick Was the problem solved?

**Comment by afreen-zemoso** [comment]
Do we have any solution for this?

**Comment by moxy-bc** [comment]
To be competitive it needs to support Bezier handles and points for curved lines, and midpoints for orthogonal lines.  This has become a standard feature on most online drawing tools.

**Comment by zhenghan2017** [comment]
@moklick Hi, Bro. I have same feature too. I finished it in my project in `11.10.2` version，Maybe i can provider it to other people. But I don't know how can i contribute code to it. If you are also interested, you can reply me.

**Reply by deepakk28** [reply]
Yes I am interested. Can you share the code with me?

**Reply by NghiaDepTrai** [reply]
Yes I am interested. Can you share the code with me?

**Reply by rootux** [reply]
Yes please

**Reply by linyuling520** [reply]
Yes I am interested. Can you share the code with me?

**Comment by Bartel-C8** [comment]
There is also a PRO example available for this now? 
https://reactflow.dev/examples/edges/editable-edge

**Comment by zh-lx** [comment]
🚀 I found a awesome example for this! It's even better than the pro example provided by official team. Please look the image:

![Jan-08-2025 11-40-29](https://github.com/user-attachments/assets/4a4beed2-e9dc-4eed-97e0-aa364caac105)


The link of example: [sandbox example](https://codesandbox.io/p/sandbox/floral-framework-st6npy)
The origin implementation ideas in [stackoverflow](https://stackoverflow.com/questions/77831116/is-it-possible-to-shape-reactflow-edges-by-dragging-them).

**Reply by moklick** [reply]
Thanks for sharing! This is great :)

**Reply by zh-lx** [reply]
> Thanks for sharing! This is great :)

❤ It would be even better if this feature could be integrated into `xyflow` internally.

**Comment by wladwnt** [comment]
There is also another example on [stackoverflow](https://stackoverflow.com/questions/77831116) provided by Daniel Cruz, I updated it to get working with the current version of reactflow. See here [https://codesandbox.io/p/sandbox/pensive-cohen-lv4hfx](https://codesandbox.io/p/sandbox/pensive-cohen-lv4hfx)

I am new to react (just few days), maybe someone can improve it further.
